src/mtev_cluster.c

```diff
@@ -680,6 +745,114 @@ mtev_cluster_do_i_own(mtev_cluster_t *c, void *key, size_t klen, int w) {
   return mtev_false;
 }
 
+mtev_boolean
+mtev_cluster_set_heartbeat_payload(mtev_cluster_t *cluster,
+    uint8_t app_id, uint8_t key, void *payload, uint8_t payload_length) {
+  assert(payload);
+
+  int i, old_idx = -1;
+  int payload_len_sum = 0;
+  hb_payload_t *hb_payload;
+
+  if(cluster->hb_payloads == NULL) {
+    cluster->hb_payloads = jl_array_list_new(free);
+  }
+
+  for(i=0; i != cluster->hb_payloads->length; ++i) {
+    hb_payload = jl_array_list_get_idx(cluster->hb_payloads, i);
+    if(hb_payload->app_id == app_id && hb_payload->key == key) {
+      old_idx = i;
+    } else {
+      payload_len_sum += hb_payload->data_len;
+    }
+  }
+
+  if(payload_len_sum + payload_length > MAX_PAYLOAD_LEN_SUM) {
+    return mtev_false;
+  }
+
+  if(old_idx != -1) {
+    hb_payload = jl_array_list_get_idx(cluster->hb_payloads, old_idx);
+  } else {
+    hb_payload = malloc(sizeof(hb_payload_t) + sizeof(payload));
```

Comment:
It's freed at line 810

please note that this is hb_payload does not store the payload itself but just a pointer to it! every time a heartbeat message is generated mtev_cluster copies the data from there to the actual message. this way the user doesn't have to inform mtev_cluster that the payload has changed.